extensions/amp-ad-network-adzerk-impl/0.1/amp-ad-network-adzerk-impl.js

```diff
@@ -0,0 +1,179 @@
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?CreativeMetaDataDef} */
+    this.creativeMetadata_ = null;
+
+    /** @private {?AmpTemplateCreativeDef} */
+    this.ampCreativeJson_ = null;
+
+    ampAdTemplates = ampAdTemplates || new AmpAdTemplates(this.win);
+  }
+
+  /**
+   * Validate the tag parameters.  If invalid, ad ad will not be displayed.
+   * @override
+   */
+  isValidElement() {
+    return !!this.getAdUrl();
+  }
+
+  /** @override */
+  getSigningServiceNames() {
+    // Does not utilize crypto signature based AMP creative validation.
+    // TODO(keithwrightbos): move import of crypto validation into
+    // implementations, reducing adzerk binary size.
+    return [];
+  }
+
+  /** @override */
+  getAdUrl() {
+    const src = this.element.getAttribute('src');
+    if (!/^https:\/\/adzerk.com\?id=\d+$/i.test(src)) {
+      return '';
+    }
+    if (getMode(this.win).localDev) {
+      return `http://ads.localhost:${this.win.location.port}` +
+        '/adzerk/' + /^https:\/\/adzerk.com\?id=(\d+)/.exec(src)[1];
+    }
+    // TODO(adzerk): specify expected src path.
+    return /^https:\/\/adzerk.com\?id=\d+$/i.test(src) ? src : '';
+  }
+
+  /** @override */
+  maybeValidateAmpCreative(bytes, headers) {
+    if (headers.get(AMP_TEMPLATED_CREATIVE_HEADER_NAME) !== 'true') {
+      return /**@type {!Promise<(ArrayBuffer|null)>}*/ (Promise.resolve(null));
+    }
+    // Shorthand for: reject promise if current promise chain is out of date.
+    const checkStillCurrent = this.verifyStillCurrent();
+    return utf8Decode(bytes).then(body => {
+      checkStillCurrent();
+      this.ampCreativeJson_ = /** @type {!AmpTemplateCreativeDef} */
+        (tryParseJson(body) || {});
+      const proxyUrl = getMode(this.win).localDev
+        ? this.ampCreativeJson_.ampCreativeTemplateUrl
+        : this.getTemplateProxyUrl_(
+            this.ampCreativeJson_.ampCreativeTemplateUrl);
+      // TODO(keithwrightbos): macro value validation?  E.g. http invalid?
+      return ampAdTemplates
+          .fetch(proxyUrl)
+          .then(parsedTemplate => {
+            this.creativeMetadata_ = /** @type {!CreativeMetaDataDef} */
+                (super.getAmpAdMetadata(parsedTemplate));
+            return utf8Encode(this.creativeMetadata_.minifiedCreative);
+          })
+          .catch(error => {
+            dev().warn(TAG, 'Error fetching/expanding template',
+                this.ampCreativeJson_, error);
+            this.forceCollapse();
+            return Promise.reject(NO_CONTENT_RESPONSE);
+          });
+    });
+  }
+
+  /**
+   * Converts the canonical template URL to the CDN proxy URL.
+   * @param {string} url
+   * @return {string}
+   */
+  getTemplateProxyUrl_(url) {
+    const loc = parseUrl(url);
+    const hostClean = startsWith(loc.host, 'www.')
+      ? loc.host.slice(4)
+      : loc.host;
+    return loc.protocol + '//' + hostClean.replace('.', '-') + '.' +
```

Comment:
please refer to [AMP cache URL format](https://developers.google.com/amp/cache/overview)

Reply:
Probably won't fix until morning, but besides removing 'www', what's wrong with this format?

Reply:
Maybe not clear from the code, but this converts `https://adserver.com/amp_template_1` to `https://adserver-com.cdn.ampproject.org/a/s/adserver.com/amp_template_1` as mentioned.

Reply:
- Converting the AMP document domain from IDN (Punycode) to UTF-8.
- Replacing every "-" (dash) with "--"(2 dashes).
- Replacing every "." (dot) with a "-" (dash).
- Converting back to IDN (Punycode).

Reply:
Done.